### PR TITLE
[WIP][MIG]school_fees: Migration to v14.

### PR DESCRIPTION
**Pull request by ForamSerpentCS**

### CODE REVIEW

File: school_fees/__manifest__.py [hunk 1]
```diff
@@ -14,10 +14,10 @@
     'data': ['security/ir.model.access.csv',
              'security/security_fees.xml',
              'views/school_fees_view.xml',
```
Comment (Trivedi-Vacha-SerpentCS) [inline school_fees/__manifest__.py]: the data file should define before view.

File: school_fees/models/school_fees.py [hunk 2]
```diff
@@ -58,21 +60,23 @@ def fees_register_confirm(self):
...
-                    raise UserError(_('There is already a Payslip exist for\
-student: %s for same date.!') % stu.name)
+                if old_slips_rec:
+                    raise ValidationError(_('''
+There is already a Payslip exist for student: %s for same date.!'''
+                    ) % stu.name)
                 else:
-                    rec.number = self.env['ir.sequence'].\
-                        next_by_code('student.fees.register') or _('New')
+                    rec.number = self.env['ir.sequence'].next_by_code(
```
Comment (Trivedi-Vacha-SerpentCS) [inline school_fees/models/school_fees.py]: self.env['ir.sequence'] define outside the for loop

File: school_fees/README.rst [hunk 3]
```diff
@@ -0,0 +1,44 @@
+school_fees
+
+This module was written to extend the functionality of fees management to support school management and allow you to manage fees through this functionality.
+
+Installation
+
+To install this module, you need to:
+
+install 'school', 'account_voucher' module
```
Comment (Trivedi-Vacha-SerpentCS) [inline school_fees/README.rst]: account_voucher module is not there in odoo14.

File: school_fees/demo/school_fees_demo.xml [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+            <field name="amount">4000.00</field>
+        </record>
+        <record id="demo_school_fees_5" model="student.payslip.line">
+            <field name="name">ICard Fees</field>
+            <field name="code">05</field>
+            <field name="type">month</field>
+            <field name="amount">20.00</field>
+        </record>
+        <record id="demo_school_fees_6" model="student.payslip.line">
+            <field name="name">Sport Fees</field>
```
Comment (Trivedi-Vacha-SerpentCS) [inline school_fees/demo/school_fees_demo.xml]: "Sport(s) Fee"

File: school_fees/models/school_fees.py [hunk 5]
```diff
@@ -0,0 +1,569 @@
...
+    _description = 'Student fees Register'
+
+    @api.depends('line_ids')
+    def _compute_total_amount(self):
+        '''Method to compute total amount'''
+        for rec in self:
+            total_amt = 0.0
+            for line in rec.line_ids:
+                total_amt += line.total
+            rec.total_amount = total_amt
```
Comment (Trivedi-Vacha-SerpentCS) [inline school_fees/models/school_fees.py:19]: Can be done in a single line using inline for loop and sum function

File: school_fees/models/school_fees.py [hunk 6]
```diff
@@ -0,0 +1,569 @@
...
+                           'currency_id':
+                           rec.company_id.currency_id.id or False,
+                           'journal_id': rec.journal_id.id,
+                           'fees_structure_id': rec.fees_structure.id or False}
+                    slip_rec = slip_obj.create(res)
+                    slip_rec.onchange_student()
+            # Calculate the amount
+            amount = 0
+            for data in rec.line_ids:
+                amount += data.total
```
Comment (Trivedi-Vacha-SerpentCS) [inline school_fees/models/school_fees.py:121]: use inline for loop

File: school_fees/models/school_fees.py [hunk 7]
```diff
@@ -0,0 +1,569 @@
...
+                         'The code of the Fees Structure must be unique !')]
+
+    @api.onchange('student_id')
+    def onchange_student(self):
+        '''Method to get standard , division , medium of student selected'''
+        self.standard_id = False
+        self.division_id = False
+        self.medium_id = False
+        if self.student_id:
+            self.standard_id = self.student_id.standard_id.id
```
Comment (Trivedi-Vacha-SerpentCS) [inline school_fees/models/school_fees.py]: Please check, remove if self.standard_id = False declartion not require.

File: school_fees/models/school_fees.py [hunk 8]
```diff
@@ -0,0 +1,569 @@
...
+        for rec in self:
+            if rec.state != 'draft':
+                raise ValidationError(_('''
+            You can delete record in unconfirm state only!'''))
+        return super(StudentPayslip, self).unlink()
+
+    @api.onchange('journal_id')
+    def onchange_journal_id(self):
+        '''Method to get currency from journal'''
+        for rec in self:
```
Comment (Trivedi-Vacha-SerpentCS) [inline school_fees/models/school_fees.py:391]: Define journal = rec.journal_id as its used multiple times in code

File: school_fees/models/school_fees.py [hunk 9]
```diff
@@ -0,0 +1,569 @@
...
+
+    def write(self, vals):
+        '''Inherited write method to update values from student model'''
+        if vals.get('student_id'):
+            student_rec = self.env['student.student'].browse(
+                                                    vals.get('student_id'))
+            vals.update({'standard_id': student_rec.standard_id.id,
+                         'division_id': student_rec.standard_id.division_id.id,
+                         'medium_id': student_rec.medium_id.id})
+        return super(StudentPayslip, self).write(vals)
```
Comment (Trivedi-Vacha-SerpentCS) [inline school_fees/models/school_fees.py:424]: Create and write method have the same block of code, Can't we code once and use twice?

File: school_fees/models/school_fees.py [hunk 10]
```diff
@@ -0,0 +1,569 @@
...
+
+    def copy(self, default=None):
+        '''Inherited copy method to assign default value at copying record'''
+        if default is None:
+            default = {}
+        default.update({'state': 'draft',
+                        'number': False,
+                        'move_id': False,
+                        'line_ids': []})
+        return super(StudentPayslip, self).copy(default)
```
Comment (Trivedi-Vacha-SerpentCS) [inline school_fees/models/school_fees.py]: Please check I guess we can eliminate this method and instead use copy=False in field definition.
the default value of the state is already defined as 'draft'

File: school_fees/models/school_fees.py [hunk 11]
```diff
@@ -0,0 +1,569 @@
...
+                             'amount': data.amount,
+                             'currency_id': data.currency_id.id or False,
+                             'currency_symbol': data.currency_symbol or False}
+                lines.append((0, 0, line_vals))
+            rec.write({'line_ids': lines})
+            # Compute amount
+            amount = 0
+            for data in rec.line_ids:
+                amount += data.amount
+            rec.register_id.write({'total_amount': rec.total})
```
Comment (Trivedi-Vacha-SerpentCS) [inline school_fees/models/school_fees.py]: Same as above, inline for loop and sum function

File: school_fees/report/student_fees_register.xml [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+<odoo>
+    <template id="student_fees_register">
+        <t t-call="web.html_container">
+            <t t-foreach="docs" t-as="o">
+                <t t-call="web.external_layout">
+                    <div class="page">
+                        <div class="oe_structure" />
+                        <div class="row" align="center">
+                            <span>
+                                <h1 class="text-center">Student Payroll Register</h1>
```
Comment (Trivedi-Vacha-SerpentCS) [inline school_fees/report/student_fees_register.xml]: Student(s) Payroll Register

File: school_fees/security/security_fees.xml [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+            <field name="name">Student Fees Record Rule</field>
+            <field name="model_id" ref="model_student_payslip"/>
+            <field name="global" eval="True"/>
+            <field name="domain_force">[('student_id.user_id','=',user.id)]</field>
+            <field name="groups" eval="[(4, ref('school.group_school_student'))]"/>
+        </record>
+
+        <!-- Record Rule Parents can see own child Fee recipt -->
+        <record id="fee_receipt_rule_parent" model="ir.rule">
+            <field name="name">Parent Fees Record Rule</field>
```
Comment (Trivedi-Vacha-SerpentCS) [inline school_fees/security/security_fees.xml]: Parent(s)

File: school_fees/security/security_fees.xml [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+            <field name="name">Parents Fees Register Rule</field>
+            <field name="model_id" ref="model_student_fees_register"/>
+            <field name="global" eval="True"/>
+            <field name="domain_force">[('standard_id','in',[stud.standard_id.standard_id.id for stud in user.partner_id.student_ids])]</field>
+            <field name="groups" eval="[(4, ref('school.group_school_parent'))]"/>
+        </record>
+        
+         <!-- Record Rule Student can see own Fee Register -->
+        <record id="fees_register_rule_student" model="ir.rule">
+            <field name="name">Student Fees Register Rule</field>
```
Comment (Trivedi-Vacha-SerpentCS) [inline school_fees/security/security_fees.xml]: Student(s)

**Review by Trivedi-Vacha-SerpentCS — CHANGES_REQUESTED** [review]

File: school_fees/__manifest__.py [hunk 15]
```diff
@@ -0,0 +1,26 @@
+# See LICENSE file for full copyright and licensing details.
+
+{
+    "name": "Fees Management",
```
Comment (JayVora-SerpentCS) [inline school_fees/__manifest__.py]: edit name

File: school_fees/report/report_student_fees_register.py [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+
+
+class ReportStudentFeesRegister(models.AbstractModel):
+    _name = "report.school_fees.student_fees_register"
+    _description = "School Fees Register Report"
+
+    def get_month(self, indate):
+        """Method to get month"""
+        out_date = indate.strftime("%B") + "-" + indate.strftime("%Y")
+        return out_date
```
Comment (Trivedi-Vacha-SerpentCS) [inline school_fees/report/report_student_fees_register.py]: No need of out_date variable, direct use 
return indate.strftime("%B") + "-" + indate.strftime("%Y")

**Review by Trivedi-Vacha-SerpentCS — CHANGES_REQUESTED** [review]

**Review by Trivedi-Vacha-SerpentCS — APPROVED** [review]

**Pull request merged**